quisp/modules/Application.cc

```diff
@@ -60,185 +24,174 @@ Application::Application() { generatePacket = nullptr; }
+
+  other_end_node_addresses = storeEndNodeAddresses();
+
+  if (!is_e2e_connection) {
+    return;
+  }
+
+  int traffic_pattern = par("TrafficPattern");
+
+  if (traffic_pattern == 0) {
+    EV_INFO << "EndToEndConnection is set true. but no traffic pattern specified; proceeding with no traffic\n";
+    return;
+  }
+
+  // just one connection
+  if (traffic_pattern == 1) {
+    int initiator_address = par("LoneInitiatorAddress");
+    if (my_address == initiator_address) {
+      int endnode_dest_addr = getOneRandomEndNodeAddress();
+      EV_INFO << "Just one lonely connection setup request will be sent from " << my_address << " to " << endnode_dest_addr << "\n";
+      ConnectionSetupRequest *pk = createConnectionSetupRequest(endnode_dest_addr, number_of_resources);
+      scheduleAt(simTime(), pk);
     }
+    return;
   }
+
+  // let's all mambo!
+  // Each EndNode makes exactly one connection.
+  // this means that some nodes will be receivers of more than one connection, at random.
+  if (traffic_pattern == 2) {
+    int endnode_dest_addr = getOneRandomEndNodeAddress();
+    EV_INFO << "My connection setup request will be sent from " << my_address << " to " << endnode_dest_addr << "\n";
+    ConnectionSetupRequest *pk = createConnectionSetupRequest(endnode_dest_addr, number_of_resources);
+    // delay to avoid conflict
+    scheduleAt(simTime() + exponential(0.00001 * my_address), pk);
+    return;
+  }
+
+  error("Invalid TrafficPattern specified.");
+}
+
+ConnectionSetupRequest *Application::createConnectionSetupRequest(int dest_addr, int num_of_required_resources) {
+  ConnectionSetupRequest *pk = new ConnectionSetupRequest();
+  pk->setActual_srcAddr(my_address);
+  pk->setActual_destAddr(dest_addr);
+  pk->setDestAddr(my_address);
+  pk->setSrcAddr(my_address);
+  pk->setNumber_of_required_Bellpairs(num_of_required_resources);
+  pk->setKind(7);
+  return pk;
 }
 
 void Application::handleMessage(cMessage *msg) {
   if (dynamic_cast<deleteThisModule *>(msg) != nullptr) {
     deleteModule();
     delete msg;
-  } else if (dynamic_cast<ConnectionSetupRequest *>(msg) != nullptr) {
-    send(msg, "toRouter");
-  } else if (dynamic_cast<ConnectionSetupResponse *>(msg) != nullptr) {
+    return;
+  }
+
+  if (dynamic_cast<ConnectionSetupRequest *>(msg) != nullptr || dynamic_cast<ConnectionSetupResponse *>(msg) != nullptr) {
     send(msg, "toRouter");
-  } else if (dynamic_cast<RejectConnectionSetupRequest *>(msg) != nullptr) {
+    return;
+  }
+
+  if (dynamic_cast<RejectConnectionSetupRequest *>(msg) != nullptr) {
     RejectConnectionSetupRequest *pk = check_and_cast<RejectConnectionSetupRequest *>(msg);
-    int actual_src = pk->getActual_srcAddr();
-    if (actual_src == myAddress) {
-      float recon_try = std::rand() / RAND_MAX;
-      int reject_node = pk->getSrcAddr();
-      EV << "Connection was rejected by " << reject_node << "at" << myAddress << "\n";
+
+    if (my_address == pk->getActual_srcAddr()) {
```

Comment:
hmm, I couldn't decide this line is right or not.
should it be destAddr or actualDestAddr?

Reply:
Mmm, I'd have to look at that to be sure.  cocori or kaaki would be a better choice to check this.

Reply:
Finally, I found this part is unreachable because the router doesn't send RejectConnectionSetupRequest packet to the Application
https://github.com/sfc-aqua/quisp/blob/master/quisp/modules/Router.cc#L131-L133

Reply:
Some of this might code I wrote, and was always incomplete. The idea was to develop this in conjunction with the Connection Setup Internet Draft, but only enough of this got implemented to make connections work under the handful of circumstances that kaaki and cocori were actually testing. Completing this part of the code would be a *very* useful task.  I think there's an open Issue somewhere...

Reply:
@rdviii did you mean [this issue #97](https://github.com/sfc-aqua/quisp/issues/97)? Handling the RejectConnectionSetupRequest must be useful. if I have enough time to complete it, I'd like to try it.
But now, this part is unreachable and I removed. Let's implement it later in other PR.

now I'm satisfied with what I asked. So I'll mark as resolved this. thanks a lot!